### ISSUE TITLE
How to express logic "and","or", "not"?

### ISSUE THREAD

**Issue by zhaopengme** [Shopify/liquid#138]
"and" is and.

"or" is or.

but what is "not"  ?

thanks!

**Comment by courtneynguyen** [comment]
I agree with Cam and AllThatIsTheCase. It is difficult to check if product.tags and customer.tags do not contain the tag "wholesale" without a "not contains" operator because of the combined condition.

**Comment by fw42** [comment]
Reopened because of https://github.com/Shopify/liquid/issues/526

**Comment by Blaisorblade** [comment]
From #526:

> same way having support for parentheses would be nice to have

Ah. Sounds like the problem is the parser support for not, because you're not using a parser generator or something like that — and you can't parse the context free grammar of expressions with parentheses with regular expressions such as https://github.com/Shopify/liquid/blob/12d526a05c0654dc67e350f170f91d62fd4875b7/lib/liquid/tags/if.rb#L14. I see why this is not on your roadmap — you'd need to replace that part of the parser.

**Comment by fw42** [comment]
@trishume and @pushrax might have thoughts

**Comment by pushrax** [comment]
We could totally implement parentheses with the new parsers, but it would probably break some weird-ass templates that are using them regardless of whether they do anything or not. We even have a test asserting that parentheses are broken https://github.com/Shopify/liquid/blob/master/test/integration/parsing_quirks_test.rb#L74 :(

**Comment by trishume** [comment]
The other option is to have some kind of super-conditional tag or syntax that is only available with liquid-c or the strict ruby parser. That way it could be opt-in and not break existing templates.

Maybe call it `cond`, `if*` or `if+`. Though the fact that fancy expression features work in that tag but not if would be weird.

Or it could be an opt in for expressions. Maybe some special kind of augmented parentheses in which real expression syntax is valid. Maybe something like the magic test command in bash: `{% if [this && (that || other)] %}`

**Comment by pushrax** [comment]
That would work, but it would also be really confusing going forward. I think if we do this at all we should work towards unifying syntax and formally specifying it.

**Comment by Blaisorblade** [comment]
As a programmer-minded user, I'd be happy to set a flag to enable an extended syntax, where my 
parentheses mean something, and promise I don't use weird templates.

One could also argue that flag should be orthogonal to strict vs lax mode. In fact, when transitioning, I'll probably need to switch to strict mode to check our templates actually parse (i.e., probably, they should contain no parentheses), before enabling a new syntax where these parenthesis suddenly can be used and have meaning.

**Comment by mikebridge** [comment]
+1 for the formal specification.  Right now there's some divergence among various flavours, e.g. the "keyword" style of [Jekyll's parameters in `include`](http://jekyllrb.com/docs/templates/#includes) vs. the ruby-variable style of [Shopify Liquid] (https://docs.shopify.com/themes/liquid-documentation/tags/theme-tags#include).

**Comment by fw42** [comment]
I'm afraid neither Shopify nor Jekyll can make any drastic changes now since a huge number of users already depends on what we/they have :-)

**Comment by mikebridge** [comment]
@fw42 Yeah, I'm thinking more of myself here Re: specs---I'm porting liquid to another environment and was hoping not to accidentally end up in the same position. :)

**Comment by Tamriel** [comment]
It's a shame this is still an issue.

**Comment by fw42** [comment]
@Tamriel: Patches welcome

**Comment by pushrax** [comment]
Well actually in this case it's not simple at all and any patch we accept is going to take a long time and a lot of testing on our end...

**Comment by rickydazla** [comment]
You're __really__ concerned that it would _"break some weird-ass templates that are using them regardless of whether they do anything or not"_?

**Comment by mikeyhew** [comment]
@pushrax why not just change the behaviour of boolean expressions to the normal way, and put that change in the next major version? And we could allow people to opt in in the current version by using a declaration at the top of the file or something.

**Comment by nickpearson** [comment]
I've used the `boolean-expressions` gem to handle boolean logic like requested, though not in the context of Liquid. I'll just leave this link here in case it helps someone who wants to submit a patch that either uses this gem (last updated in 2012) or extracts its functionality. https://github.com/meh/boolean-expression

It's not immediately clear how to use it from the README, so see the [`#evaluate` tests](https://github.com/meh/boolean-expression/blob/master/test/expression_spec.rb#L42) for more examples. The basic gist is `Boolean::Expression[<expression-string>][<symbols that are true>]`, like `Boolean::Expression['a && b'][:a, :b]` (which returns true, because both `:a` and `:b` are passed as `true` values).

I agree that this is a big change to Liquid that will require tons of testing, but it would be nice to be able to write more complex `if` statements, so maybe this will help someone.

**Comment by Blaisorblade** [comment]
> And we could allow people to opt in in the current version by using a declaration at the top of the file or something.

Given what the compatibility requirements seem, maybe it'd be more realistic to require people to opt-in to the new syntax. I started sketching details, then realized [I already did to some extent](https://github.com/Shopify/liquid/issues/138#issuecomment-75424662).

The issue is recomposing conflicting requirements. @pushrax had volunteered to do the work, but [added](https://github.com/Shopify/liquid/issues/138#issuecomment-75414507):
> [Opt-in new syntax] would work, but it would also be really confusing going forward. I think if we do this at all we should work towards unifying syntax and formally specifying it.

while @fw42 [replied](https://github.com/Shopify/liquid/issues/138#issuecomment-87851336):
> I'm afraid neither Shopify nor Jekyll can make any drastic changes now since a huge number of users already depends on what we/they have :-)

Options include:

1. go for an opt-in better syntax: ugly but better than now. Add an API call to enable that globally for who wants it. Possibly, reuse Python's `future` solution, so that you declare in advance some future major version will enable this by default, in case you want the API to get eventually sane. Or wait for the whole community to test the new syntax before deciding what to do.
2. close the issue as won't fix.
3. change syntax and check for breakage (so complicated that it didn't happen in one year)
4. change syntax and ignore breakage (no)

@pushrax, @fw42, is option (1) doable on your end? The only disadvantage seems "ugliness" — I care about software's elegance, because it makes a practical difference, but it seriously seems less ugly than (2). Are there further obstacles? Thoughts?

**Comment by evulse** [comment]
I think this issue needs to be broken in to 2 parts.

1. Some of the issues are about not having a NOT operator
2. Adding support for parentheses to handle order of operation

I think with any problem its better to start with the easy wins so I have just committed code around the first issue to hopefully get it out of the way so that only the second issue remains which from discussion is the one that is going to cause the biggest impact on backwards compatibility. Also It requires a lot more complexity by needing to build an AST with nodes and then resolving them back in order.

Comments on the Pull Request are appreciated.

**Comment by evulse** [comment]
This simplifies the following 

````
{% if a and b %}
  {% unless c %}
    ...
  {% endunless %}
{% endif %}
````

to be a single line

````
{% if a and b and not c %}
    ....
{% endif %}
````

It does not support the following as it is already available with unless

````
{% if not a and b and not c %}
    ....
{% endif %}
````

Already existing unless

````
{% unless a and b and not c %}
    ....
{% endif %}
````

**Comment by Blaisorblade** [comment]
To be clear: your patch does not support `not` at the beginning of the expression, right?

Are your two last examples supposed to be equivalent? If so I don't get the precedence rules or something. I infer from your patch that `not` binds most tightly.

Pretending we have parentheses, `unless a and b and not c` should mean `unless (a and b and (not c))`, that is `if not (a and b and (not c))`, that is `if (not a) or (not b) or c`, with `or` instead of `and`.

**Comment by TWiStErRob** [comment]
I get the first code example, but the second is weird. Are you saying there can be only one `not` and only as the last one? Also in common Boolean logic:
```
!a and b and !c <> !(a and b and !c) === !a or !b or c
```
I'm using parentheses above because as far as I know `unless` negates the condition altogether.

**Comment by Blaisorblade** [comment]
@TWiStErRob the patch supports `if a or not b or not c`, because it parses `and not` and `or not`, but doesn't support `not` on its own.

On the example I agree, we commented simultaneously 😄

**Comment by evulse** [comment]
unless doesn't operate like you think it would. It only affects the first expression and then reverts to a normal if like below

````
{% unless a and b and c %)
````

is actually
````
!a && b && c
````

**Comment by evulse** [comment]
I might be wrong I'm just referencing the codebase

> # After the first condition unless works just like if

**Comment by TWiStErRob** [comment]
So if I want to say `if not a or b` (simple Boolean implication) then I have to write `if b or not a` (reversed) or `unless a and not b` (negated + DeMorgan)?

**Comment by evulse** [comment]
With the current pull request. Its not a big change to allow not at the beginning as well.

**Comment by TWiStErRob** [comment]
> unless doesn't operate like you think it would

Have you heard of the  principle of least astonishment?

> unless: Like if, but executes a block of code only if a certain condition is not met (that is, if the result is false). -- https://help.shopify.com/themes/liquid/tags/control-flow-tags

**Comment by Blaisorblade** [comment]
@TWiStErRob The pull request does not change the behavior of `unless`.
But yes, the behavior he describes is not sane. Yet, looking at this implementation of `unless`, I fear @evulse is right, though it might be good to try this out:
https://github.com/evulse/liquid/blob/e07529dbfa1bfbe908d8dfb1605274cb0b4c3436/lib/liquid/tags/unless.rb

BTW, remember: nobody is saying the current code makes sense, just that there are terrible backward compatibility constraints.
https://github.com/Shopify/liquid/issues/138#issuecomment-75179995

**Comment by TWiStErRob** [comment]
Phew, @evulse read it wrong, a "block" is stuff between start and end tags of liquid. It's possible to write:
```liquid
{% unless 'a' == 'a' %} hello {% elsif 'b' != 'c' %} cruel {% else %} world {% endunless %}
```
which outputs "cruel". I confirmed this by printing the `@blocks` variable in `unless.rb`:
```
[#<Condition 'a' == 'a'>, #<Condition 'b' != 'c'>, #<Condition >]
```

**Comment by Blaisorblade** [comment]
@TWiStErRob Thanks for trying this out and clearing the misunderstanding! Now we can put down the pitchforks and relax again 😄

**Comment by TWiStErRob** [comment]
I think if this change is made, then it should accept `not` right after `if`/`unless` too, again referencing the principle of least astonishment, mostly because currently (live release 3.0.6) this is accepted code:
```
{% unless not 'a' == 'a' or not 'b' %} hello {% elsif not 'b' != 'c' %} cruel {% else not %} world {% endunless %}
```
even though the result is undefined, I guess... I like `else not` the best! :)

I now better understand @pushrax's concerns about breaking people's code. I bet if we just simply enabled a strict(er) syntax checker/parser many people's code would simply break with trivial fixes. Is there such an option now?

**Comment by evulse** [comment]
New and Improved version. Now supports NOT as standalone so it can be used before the first expression as well.

So ignore all previous comments.

**Comment by nickpearson** [comment]
@evulse check out the [boolean-expression gem](https://github.com/meh/boolean-expression). It may save you some work on parsing and building an AST when you add support for parentheses. It also handles "not" although it uses `!` for this.

I use this in one of my apps, and it works great. The only thing that wasn't immediately intuitive when I first started writing code was that the gem only expects things that are already booleans and just does the boolean logic for you. In other words, it expects `a && b`, not `a == 1 && b == 2`. So, to evaluate the latter expression, you'd use something like this:

```ruby
# just setting up vars for demonstration
a = 1
b = 2

# build an array to pass which parts of the expression are true
bools = []
bools << :a if a == 1
bools << :b if b == 2

# evaluate the expression
Boolean::Expression['a && b'][*bools] #=> true
```

See the [spec](https://github.com/meh/boolean-expression/blob/master/test/expression_spec.rb) for some more example usage. (The README is sparse.)

I know adding a gem dependency may not be ideal or even possible, but perhaps some of the code from boolean-expression can be used to save you the headache of implementing this behavior on your own.

**Comment by evulse** [comment]
@nickpearson thanks I did see your previous comment and had a look at that one and one other library and their logic is all very similar so will definitely benefit from their approach. The hard part with implementation is that the booleans aren't calculated until right at the end when the template is actually being rendered. So ultimately we need a hybrid of both. Layering the liquid expression classes within the structure of those libraries so at render time it has all the information it needs to evaluate the booleans and process them efficiently.

Any suggestions on an approach that could work without restructuring the current condition class would be helpful

**Comment by nickpearson** [comment]
I _thought_ I remembered commenting about that gem somewhere! I completely forgot it was here -- sorry for the double post. :)

I'll think more about the structuring you mentioned and will post here if I come up with a better way to handle this, but here are my thoughts after a cursory glance.

Unless I'm making this too complicated, I wonder if it would make sense to replace [`Condition`](https://github.com/Shopify/liquid/blob/master/lib/liquid/condition.rb) with a new `Conditions` class, with `If` simply passing the unparsed string of conditions (everything between `{% if` and `%}`). For example, pass `(a == 1 or a == 2) and b != 3` to `Conditions` to be parsed and evaluated in the same class, where during render it would first evaluate `a == 1`, then `a == 2`, then `b != 3`, and then evaluate them all as a single boolean expression: `(x || y) && z`.

Performance-wise, this can cause extra variable lookups and comparisons, because it has to perform every comparison in order to hand all the `true`/`false` values to the boolean expression evaluator. In my example, that means `b` will be looked up and compared to `3` even if `a` is neither `1` nor `2`. This could be avoided by passing procs to the expression evaluator so that it requests each variable only as needed, so that e.g. `z` would not be requested if `x` and `y` are both `false`. (The proc behavior does not exist yet to my knowledge.)

**Comment by evulse** [comment]
Just thought I would add what the current NOT feature would break on existing templates so the only case I can think of is

````
{% assign not = true %}
{% if not == true %}
Pass
{% endif %}
````

This is because not has become a reserved word however I would be able to keep backwards compatibility if I looked forward first and checked if it was followed by a operator( ==, !=) etc.

Thoughts and does anyone else have any edge cases we need to consider?

**Comment by evulse** [comment]
Keeping support for the above is going to be nasty. I honestly recommend doing the following as it needs to happen at some point

1. Declare a new major version
2. Define strict as the default error_mode
3. Remove lax as an option, but still allow the current warn mode
3. Declare the lax parser as deprecated and it will no longer receive any changes
4. Anyone using v3 who wants to transition to v4 should use the warn mode.
5. Recommend that anyone using the warn mode should set up functionality to identify users who are receiving SyntaxErrors and setup a notification system that a transition is happening and they must resolve their issues as v5 will no longer support the lax parser at all.

**Comment by evulse** [comment]
@fw42 @dylanahsmith @tobi @parkr thoughts on the above? I would love to commit the following to v4

1. Full parenthesis support to the strict parser
2. Full NOT support for the strict parser
3. Bring AND, OR and NOT into the lexer rather than the parser treating them as :id
4. Introduce the &&, II and ! syntax as well as its no additional work and these are not being used by any other functions.

Reasoning for the above is to maintain the current simplicity for the designers but obviously add the complexity that developers working on larger projects need and are accustomed to.

Also the other big thing missing from liquid which would love to introduce is array assignment. I actually think just adding support for json filter would be the most universal way to handle this. We currently have to support really odd data structures in metafields which could be simplified a lot by storing it in a single metafield as json.

**Comment by fw42** [comment]
@boourns @Thibaut any thoughts?

**Comment by Thibaut** [comment]
> the only case I can think of is
>
> ```
> {% assign not = true %}
> {% if not == true %}
> Pass
> {% endif %}
> ```

No one is doing that in Shopify, so potentially this could be an acceptable "breaking" change.

I don't think Liquid should have two different syntax, at least not in Shopify. We would need to make all our customer's templates strict-valid before considering deprecating the lax parser. 

cc @davidcornu

**Comment by ubershmekel** [comment]
Is `not` happening?

**Comment by jmarca** [comment]
I just started using Liquid in a Jekyll project, and I immediately hit the "can't use proper logic in an if statement" bug.  I got around it by faking a dummy variable using 

```
{% assign not_last_crumb = 1 %}
{% if include.last_crumb_url %}
    {% assign not_last_crumb = nil %}
{% endif %}
...
{% if forloop.last and not_last_crumb %} ... {% endif %}
```

but it would be so much cleaner to just say `if ... and not include.last_crumb_url`

After reading this bug report and seeing that it is 6 years old, and apparently a fix has been sitting here for two years, I'm kind of at a loss.  

I respectfully request that this be made a priority and fixed.  But I understand you don't want to break your existing customers at shopify, and verifying "not breaking" is likely a painful job.

**Comment by nickpearson** [comment]
Agree that this would be nice and make things simpler, but in many cases you can accomplish the same thing by comparing against a value, or using the built-in `blank` and `empty`. I'm not familiar with Jekyll, but if I understand your example correctly, it could be written as:

`{% if forloop.last and include.last_crumb_url != blank %} ... {% endif %}`

And if you're checking against a boolean variable, you can simply use `!= true` for the `not` case:

`{% if some_var and another_var != true %} ... {% endif %}`

`!= false` works as well, as would `!= nil` and `!= null`.

**Comment by jmarca** [comment]
Yeah there are plenty of work arounds…but all I'm saying is if this patch is just sitting here waiting, the work arounds shouldn't be necessary.

**Comment by aklef** [comment]
Bump because I ran into this issue this afternoon and had to waste a couple hours with a workaround..

A "not" operator would be _really_ nice

**Comment by haselwarter** [comment]
lol

**Comment by fer-rum** [comment]
Somehow, a template language that has been unable (or unwilling) to implement `not` in nearly 8 years is the foundation of several businesses and nearly all pages hosted on _github_ and _gitlab_. :astonished:

**Comment by tobiasvl** [comment]
I also ran into this issue just now (not with Shopify, but GitHub Pages) and spent some time figuring out that `not` isn't supported. GitHub unhelpfully just told me that the build failed, not why.

It's really quite astonishing that there's no `not` operator, and it doesn't seem to be mentioned in the Liquid docs either. Its absence is a major gotcha for anyone remotely familiar with programming, so a section about this issue in [the section about truthiness/falsiness](https://shopify.dev/docs/themes/liquid/reference/basics/true-and-false) or something would probably be nice. If you're not planning on adding support for it, that is.

**Comment by fer-rum** [comment]
In case somebody stumble over this in the future:

Since `unless` is basically the way of expressing _if not_ and nested loops are an implicit _and_ you can express
`if a and not b` as

```liquid
{% if a %}
    {% unless b %}
        <p> a and not b </p>
    {% endunless %}
{% endif %}
```

Sadly there seems to be no good way to express an `or` so you will have to apply [DeMorgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws). 

For more complex expressions I would recommend to use intermediate variables and evaluate the sub-expressions step by step. This is pretty tedious however and hard to maintain.

**Comment by pigsflew** [comment]
I don't know if this is really the same issue, but i'd love a not-where clause for filtering lists. IE: `{% assign promos = site.pages | where: 'homepage', true | except: 'show', true %}`.

That way anything that *isn't* `true` will trigger the thing, even if it's not actually a 'falsey' value.

**Comment by rianfloo** [comment]
Any news 9 years later? :(

**Comment by fer-rum** [comment]
Maybe they want to keep it around as a running gag. :clown_face:

**Comment by TWiStErRob** [comment]
@erikplatte `{% unless a=12 %}` exists for that reason... the problem in this issues is more complex boolean expressions, e.g. `a and not b`.

**Comment by tim-harding** [comment]
Is this on the table 11 years later? Would love a simpler way to negate an assignment than this:

```liquid
assign negated = true
if myvar
  assign negated = false
endif
```

**Comment by isaacbowen** [comment]
miiight be a thing in liquid 6.0.0? see #2017, which appears to be sketching out that future
